test/test_forms.jl

```diff
@@ -0,0 +1,31 @@
+using ITensors
+using Graphs
+using NamedGraphs
+using ITensorNetworks
+using ITensorNetworks: delta_network, bra_ket_vertices, gradient
+using Test
+using Random
+
+@testset "Forms" begin
+  g = named_grid((1, 4))
+  s = siteinds("S=1/2", g)
+  s_operator = union_all_inds(s, prime(s; links=[]))
+  χ = 2
+  Random.seed!(1234)
+  ψket = randomITensorNetwork(s; link_space=χ)
+  ψbra = randomITensorNetwork(s; link_space=χ)
+  A = delta_network(s_operator)
+
+  blf = BilinearForm(ψbra, A, ψket)
```

Comment:
Maybe the constructor should be `BilinearForm(ψbra, ψket, A)`? Say following a convention that the state (active vertices) go first, then the operator (passive vertices).

Reply:
Another option is `BilinearForm(A, bra, ket)` and `QuadraticForm(A, ket)`, i.e. the operator is passed first. That might make more sense since the operator is more baked into the object, while the states are likely to be modified.

A slightly more exotice interface could be `BilinearForm(A)(bra, ket)` and `QuadraticForm(A)(ket)`, i.e. `BilinearForm(A)` and `QuadraticForm(A)` create functions that takes states, which is closer to the math definitions. Then `BilinearForm()(bra, ket)` and `QuadraticForm()(ket)` could make forms with identity operators.

Reply:
This leads me to think that we should actually call these types `AbstractFormNetwork`, `BilinearFormNetwork`, and `QuadraticFormNetwork`, since I think they are more specialized than simple bilinear forms and quadratic forms (i.e. they actually define a whole set of bilinear forms and quadratic forms, by removing certain subsets of bra and ket vertices). That opens up the possibility of defining `AbstractForm`, `BilinearForm`, and `QuadraticForm` types which are closer to the original definitions and not particular to wrapping a tensor network.